design-proposals/instancetype.kubevirt.io/support-declarative-vm-management.md

```diff
@@ -0,0 +1,277 @@
+
+## LiveUpdate support
+
+Users will be able to invoke the existing LiveUpdate workflow by patching the `spec.{instancetype,preference}`.
+
+The VirtualMachine controller will attempt to align the status with the spec and in doing so will invoke the required logic to attempt a LiveUpdate that may result in the hot plug of additional supported resources etc.
+
+The same will also be possible when moving between generations of an instance type if the new generation introduces additional supported resources such as vCPUs exposed as sockets or memory.
+
+## API Examples
+
+For a simple VirtualMachine using a VirtualMachineClusterInstancetype named `foo` the following details should end up in status:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When moving between instance types users will now only need to modify the name (and kind if originally provided) with the VirtualMachine controller then attempting to reconcile the spec with the status of the VirtualMachine:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+A user patches `spec.instancetype.name` with `fooNew`, `foo` is still referenced by `status.instancetypeRef.name`:
+
+```yaml
+spec:
+  instancetype:
+    name: fooNew
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+Eventually the VirtualMachine controller aligns `status.instancetypeRef.name` with `spec.instancetype.name`:
+
+```yaml
+spec:
+  instancetype:
+    name: fooNew
+[..]
+status
+  instancetypeRef:
+    name: fooNew
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When `kind` and/or `revisionName` are already populated in existing VirtualMachines these values are mirrored into status by the controller:
+
+```yaml
+spec:
+  instancetype:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    revisionName: bar
+[..]
+status
+  instancetypeRef:
+    name: foo
+    kind: virtualmachineclusterinstancetype
+    controllerRevisionRef:
+      name: bar
+```
+
+When using `inferFromVolume` with a new VirtualMachine the Matcher within the spec is no longer mutated with all runtime derived details now stashed in status:
+
+```yaml
+spec:
+  instancetype:
+    inferFromVolume: foo
+status
+  instancetypeRef:
+    name: derivedName
+    kind: derivedKind
+    controllerRevisionRef:
+      name: derivedRevisionName
+
+```
+
+If a snapshot is taken of this VirtualMachine however the values from status will be written back into the spec:
+
+```yaml
+spec:
+  instancetype:
+    name: derivedName
+    kind: derivedKind
+    inferFromVolume: foo
```

Comment:
Right now it is not valid to have `name`, `kind` and `inferFromVolume`. Maybe better drop `inferFromVolume` in this case?

Reply:
ACK I was thinking it would be better to keep the link to the original volume here but I think you're right on reflection.